Avoid conflicts with other cobra auto completions

dtaniwaki opened the pull request:
**What type of PR is this?**

/kind bug

**What this PR does / why we need it**:

`kubectl get -n foo pod [TAB][TAB]` ignores the namespace while `kubectl get --namespace foo pod [TAB][TAB]` and `kubectl get --namespace=foo pod [TAB][TAB]` don't.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:

N/A

**Special notes for your reviewer**:

N/A

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

idealhack commented:
/ok-to-test

idealhack commented:
I don't think I can reproduce this bug

```
Client Version: version.Info{Major:"1", Minor:"12", GitVersion:"v1.12.0", GitCommit:"0ed33881dc4355495f623c6f22e7dd0b7632b7c0", GitTreeState:"clean", BuildDate:"2018-09-28T15:18:13Z", GoVersion:"go1.11", Compiler:"gc", Platform:"darwin/amd64"}
```

dtaniwaki commented:
Here's my result.

```
$ kubectl version
Client Version: version.Info{Major:"1", Minor:"12", GitVersion:"v1.12.2", GitCommit:"17c77c7898218073f14c8d573582e8d2313dc740", GitTreeState:"clean", BuildDate:"2018-10-30T21:39:38Z", GoVersion:"go1.11.1", Compiler:"gc", Platform:"darwin/amd64"}

$ kubectl get -n kube-system pod
my-pod-1
my-pod-2

$ kubectl get --namespace kube-system pod
coredns-99b9bb8bd-fpwfr
etcd-minikube
kube-addon-manager-minikube
kube-apiserver-minikube
kube-controller-manager-minikube
kube-proxy-mkl4h
kube-scheduler-minikube
kubernetes-dashboard-7db4dc666b-vbnhv
storage-provisioner
tiller-deploy-64c9d747bd-6r7cn
```

dtaniwaki commented:
/assign @deads2k

dtaniwaki commented:
@deads2k Could you review this PR?

liggitt commented on pkg/kubectl/cmd/cmd.go:
```diff
@@ -84,7 +84,11 @@ __kubectl_override_flags()
     for w in "${words[@]}"; do
         if [ -n "${two_word_of}" ]; then
-            eval "${two_word_of##*-}=\"${two_word_of}=\${w}\""
+            if [[ "${two_word_of}" == --* ]]; then
```
The bug is in the autocompletion… the suggestions on [tab][tab] do not honor the specified namespace

dtaniwaki replied:
> The bug is in the autocompletion… the suggestions on [tab][tab] do not honor the specified namespace
Yes, that is what I wanted to fix, but it seems I was looking into a wrong place. I'm trying to update reproducible steps and fix the issue.

dtaniwaki commented:
I added debug logs for auto completion commands and found which patterns don't work correctly.

✅ `kubectl get pod -n argo`
✅ `kubectl get pod --namespace=argo`
❌ `kubectl get pod --namespace argo`
✅ `kubectl get -n argo pod`
✅ `kubectl get --namespace=argo pod`
✅ `kubectl get --namespace argo pod`

I enabled debug logs and got the following.

<details><summary>the completion log</summary>

```
__kubectl_handle_word: c is 0 words[c] is kubectl
__kubectl_handle_command: c is 0 words[c] is kubectl
__kubectl_handle_command: looking for _kubectl_root_command
__kubectl_handle_word: c is 1 words[c] is get
__kubectl_handle_command: c is 1 words[c] is get
__kubectl_handle_command: looking for _kubectl_get
__kubectl_handle_word: c is 2 words[c] is pod
__kubectl_handle_noun: c is 2 words[c] is pod
__kubectl_handle_word: c is 3 words[c] is -n
__kubectl_handle_flag: c is 3 words[c] is -n
__kubectl_handle_flag: looking for -n
__kubectl_handle_flag: found a flag -n, skip the next argument
__kubectl_handle_reply
__kubectl_parse_get: get completion by kubectl get -n=argo -o template --template="{{ range .items  }}{{ .metadata.name }} {{ end }}" "pod"
__kubectl_handle_word: c is 0 words[c] is kubectl
__kubectl_handle_command: c is 0 words[c] is kubectl
__kubectl_handle_command: looking for _kubectl_root_command
__kubectl_handle_word: c is 1 words[c] is get
__kubectl_handle_command: c is 1 words[c] is get
__kubectl_handle_command: looking for _kubectl_get
__kubectl_handle_word: c is 2 words[c] is pod
__kubectl_handle_noun: c is 2 words[c] is pod
__kubectl_handle_word: c is 3 words[c] is -n
__kubectl_handle_flag: c is 3 words[c] is -n
__kubectl_handle_flag: looking for -n
__kubectl_handle_flag: found a flag -n, skip the next argument
__kubectl_handle_reply
__kubectl_parse_get: get completion by kubectl get -n=argo -o template --template="{{ range .items  }}{{ .metadata.name }} {{ end }}" "pod"
__kubectl_handle_word: c is 0 words[c] is kubectl
__kubectl_handle_command: c is 0 words[c] is kubectl
__kubectl_handle_command: looking for _kubectl_root_command
__kubectl_handle_word: c is 1 words[c] is get
__kubectl_handle_command: c is 1 words[c] is get
__kubectl_handle_command: looking for _kubectl_get
__kubectl_handle_word: c is 2 words[c] is pod
__kubectl_handle_noun: c is 2 words[c] is pod
__kubectl_handle_word: c is 3 words[c] is --namespace=argo
__kubectl_handle_flag: c is 3 words[c] is --namespace=argo
__kubectl_handle_flag: looking for --namespace=
__kubectl_handle_reply
__kubectl_parse_get: get completion by kubectl get --namespace=argo -o template --template="{{ range .items  }}{{ .metadata.name }} {{ end }}" "pod"
__kubectl_handle_word: c is 0 words[c] is kubectl
__kubectl_handle_command: c is 0 words[c] is kubectl
__kubectl_handle_command: looking for _kubectl_root_command
__kubectl_handle_word: c is 1 words[c] is get
__kubectl_handle_command: c is 1 words[c] is get
__kubectl_handle_command: looking for _kubectl_get
__kubectl_handle_word: c is 2 words[c] is pod
__kubectl_handle_noun: c is 2 words[c] is pod
__kubectl_handle_word: c is 3 words[c] is --namespace=argo
__kubectl_handle_flag: c is 3 words[c] is --namespace=argo
__kubectl_handle_flag: looking for --namespace=
__kubectl_handle_reply
__kubectl_parse_get: get completion by kubectl get --namespace=argo -o template --template="{{ range .items  }}{{ .metadata.name }} {{ end }}" "pod"
__kubectl_handle_word: c is 0 words[c] is kubectl
__kubectl_handle_command: c is 0 words[c] is kubectl
__kubectl_handle_command: looking for _kubectl_root_command
__kubectl_handle_word: c is 1 words[c] is get
__kubectl_handle_command: c is 1 words[c] is get
__kubectl_handle_command: looking for _kubectl_get
__kubectl_handle_word: c is 2 words[c] is pod
__kubectl_handle_noun: c is 2 words[c] is pod
__kubectl_handle_word: c is 3 words[c] is --namespace
__kubectl_handle_flag: c is 3 words[c] is --namespace
__kubectl_handle_flag: looking for --namespace
__kubectl_handle_word: c is 4 words[c] is argo
__kubectl_handle_noun: c is 4 words[c] is argo
__kubectl_handle_reply
__kubectl_parse_get: get completion by kubectl get --namespace=argo -o template --template="{{ range .items  }}{{ .metadata.name }} {{ end }}" "argo"
__kubectl_handle_word: c is 0 words[c] is kubectl
__kubectl_handle_command: c is 0 words[c] is kubectl
__kubectl_handle_command: looking for _kubectl_root_command
__kubectl_handle_word: c is 1 words[c] is get
__kubectl_handle_command: c is 1 words[c] is get
__kubectl_handle_command: looking for _kubectl_get
__kubectl_handle_word: c is 2 words[c] is pod
__kubectl_handle_noun: c is 2 words[c] is pod
__kubectl_handle_word: c is 3 words[c] is --namespace
__kubectl_handle_flag: c is 3 words[c] is --namespace
__kubectl_handle_flag: looking for --namespace
__kubectl_handle_word: c is 4 words[c] is argo
__kubectl_handle_noun: c is 4 words[c] is argo
__kubectl_handle_reply
__kubectl_parse_get: get completion by kubectl get --namespace=argo -o template --template="{{ range .items  }}{{ .metadata.name }} {{ end }}" "argo"
__kubectl_handle_word: c is 0 words[c] is kubectl
__kubectl_handle_command: c is 0 words[c] is kubectl
__kubectl_handle_command: looking for _kubectl_root_command
__kubectl_handle_word: c is 1 words[c] is get
__kubectl_handle_command: c is 1 words[c] is get
__kubectl_handle_command: looking for _kubectl_get
__kubectl_handle_word: c is 2 words[c] is -n
__kubectl_handle_flag: c is 2 words[c] is -n
__kubectl_handle_flag: looking for -n
__kubectl_handle_flag: found a flag -n, skip the next argument
__kubectl_handle_word: c is 4 words[c] is pod
__kubectl_handle_noun: c is 4 words[c] is pod
__kubectl_handle_reply
__kubectl_parse_get: get completion by kubectl get -n=argo -o template --template="{{ range .items  }}{{ .metadata.name }} {{ end }}" "pod"
__kubectl_handle_word: c is 0 words[c] is kubectl
__kubectl_handle_command: c is 0 words[c] is kubectl
__kubectl_handle_command: looking for _kubectl_root_command
__kubectl_handle_word: c is 1 words[c] is get
__kubectl_handle_command: c is 1 words[c] is get
__kubectl_handle_command: looking for _kubectl_get
__kubectl_handle_word: c is 2 words[c] is -n
__kubectl_handle_flag: c is 2 words[c] is -n
__kubectl_handle_flag: looking for -n
__kubectl_handle_flag: found a flag -n, skip the next argument
__kubectl_handle_word: c is 4 words[c] is pod
__kubectl_handle_noun: c is 4 words[c] is pod
__kubectl_handle_reply
__kubectl_parse_get: get completion by kubectl get -n=argo -o template --template="{{ range .items  }}{{ .metadata.name }} {{ end }}" "pod"
__kubectl_handle_word: c is 0 words[c] is kubectl
__kubectl_handle_command: c is 0 words[c] is kubectl
__kubectl_handle_command: looking for _kubectl_root_command
__kubectl_handle_word: c is 1 words[c] is get
__kubectl_handle_command: c is 1 words[c] is get
__kubectl_handle_command: looking for _kubectl_get
__kubectl_handle_word: c is 2 words[c] is --namespace=argo
__kubectl_handle_flag: c is 2 words[c] is --namespace=argo
__kubectl_handle_flag: looking for --namespace=
__kubectl_handle_word: c is 3 words[c] is pod
__kubectl_handle_noun: c is 3 words[c] is pod
__kubectl_handle_reply
__kubectl_parse_get: get completion by kubectl get --namespace=argo -o template --template="{{ range .items  }}{{ .metadata.name }} {{ end }}" "pod"
__kubectl_handle_word: c is 0 words[c] is kubectl
__kubectl_handle_command: c is 0 words[c] is kubectl
__kubectl_handle_command: looking for _kubectl_root_command
__kubectl_handle_word: c is 1 words[c] is get
__kubectl_handle_command: c is 1 words[c] is get
__kubectl_handle_command: looking for _kubectl_get
__kubectl_handle_word: c is 2 words[c] is --namespace=argo
__kubectl_handle_flag: c is 2 words[c] is --namespace=argo
__kubectl_handle_flag: looking for --namespace=
__kubectl_handle_word: c is 3 words[c] is pod
__kubectl_handle_noun: c is 3 words[c] is pod
__kubectl_handle_reply
__kubectl_parse_get: get completion by kubectl get --namespace=argo -o template --template="{{ range .items  }}{{ .metadata.name }} {{ end }}" "pod"
__kubectl_handle_word: c is 0 words[c] is kubectl
__kubectl_handle_command: c is 0 words[c] is kubectl
__kubectl_handle_command: looking for _kubectl_root_command
__kubectl_handle_word: c is 1 words[c] is get
__kubectl_handle_command: c is 1 words[c] is get
__kubectl_handle_command: looking for _kubectl_get
__kubectl_handle_word: c is 2 words[c] is --namespace
__kubectl_handle_flag: c is 2 words[c] is --namespace
__kubectl_handle_flag: looking for --namespace
__kubectl_handle_word: c is 3 words[c] is argo
__kubectl_handle_noun: c is 3 words[c] is argo
__kubectl_handle_word: c is 4 words[c] is pod
__kubectl_handle_noun: c is 4 words[c] is pod
__kubectl_handle_reply
__kubectl_parse_get: get completion by kubectl get --namespace=argo -o template --template="{{ range .items  }}{{ .metadata.name }} {{ end }}" "pod"
__kubectl_handle_word: c is 0 words[c] is kubectl
__kubectl_handle_command: c is 0 words[c] is kubectl
__kubectl_handle_command: looking for _kubectl_root_command
__kubectl_handle_word: c is 1 words[c] is get
__kubectl_handle_command: c is 1 words[c] is get
__kubectl_handle_command: looking for _kubectl_get
__kubectl_handle_word: c is 2 words[c] is --namespace
__kubectl_handle_flag: c is 2 words[c] is --namespace
__kubectl_handle_flag: looking for --namespace
__kubectl_handle_word: c is 3 words[c] is argo
__kubectl_handle_noun: c is 3 words[c] is argo
__kubectl_handle_word: c is 4 words[c] is pod
__kubectl_handle_noun: c is 4 words[c] is pod
__kubectl_handle_reply
__kubectl_parse_get: get completion by kubectl get --namespace=argo -o template --template="{{ range .items  }}{{ .metadata.name }} {{ end }}" "pod"
```

</details>

I found the completion script has the following lines.

```bash
...
    # skip the argument to a two word flag
    if __kubectl_contains_word "${words[c]}" "${two_word_flags[@]}"; then
        c=$((c+1))
        # if we are looking for a flags value, don't show commands
        if [[ $c -eq $cword ]]; then
            commands=()
        fi
    fi
...
```

This code skips next argument of two word flags but it seems only `-n` is recognized as a two word flag while `--namespace` is not.

```bash
...
    flags+=("--namespace=")
    flags_with_completion+=("--namespace")
    flags_completion+=("__kubectl_get_resource_namespace")
    two_word_flags+=("-n")
    flags_with_completion+=("-n")
    flags_completion+=("__kubectl_get_resource_namespace")
...
```

And I dug into the upstream code and found that part of the script is generated by cobra.

So I fixed it and sent a PR to cobra.
https://github.com/spf13/cobra/pull/807

We can fix this issue by using a cobra version with this fix.

I actually have sent a PR to update the Cobra version which addresses another issue but doesn't include this fix.
https://github.com/kubernetes/kubernetes/pull/71447

I will fix either of the PRs depending on when new Cobra version will be released.

dtaniwaki committed:
Avoid conflicts with other cobra auto completion

dtaniwaki commented:
I updated the PR as spf13/cobra#807 got merged. I also confirmed this change fixes the completion failure of `kubectl get pod --namespace argo`.

This PR also fixes the custom func name problem of #71447.

@liggitt @dixudx @adohe @rootfs Could you review the PR?

dtaniwaki commented:
/assign @apelisse

apelisse commented:
/assign @sttts

dtaniwaki commented:
@sttts Could you review this PR?

dixudx commented on pkg/kubectl/cmd/cmd.go:
```diff
@@ -126,7 +126,9 @@ __kubectl_parse_config()
     local template kubectl_out
     template="{{ range .$1  }}{{ .name }} {{ end }}"
-    if kubectl_out=$(kubectl config $(__kubectl_override_flags) -o template --template="${template}" view 2>/dev/null); then
+    completion_cmd="kubectl config $(__kubectl_override_flags) -o template --template=\"${template}\" view"
```
Could we have this into a new function, like `__kubectl_debug_out`?

dtaniwaki replied:
Let me try!

dixudx replied:
Looks great now. Awesome!

dtaniwaki commented:
@dixudx Did it! Could you review it again?

dixudx commented:
/retest

dtaniwaki commented:
@dixudx I'm not sure if the errors came from my change. Could you take a look of it?

dtaniwaki commented:
FYI. I did the following steps to update cobra.

```bash
$ ./hack/pin-dependency.sh github.com/spf13/cobra 67fc4837d267bc9bfd6e47f77783fcc3dffc68de
$ ./hack/update-vendor.sh
```

Please tell me if it's a wrong way.

dixudx commented:
> FYI. I did the following steps to update cobra.
> 
> ```shell
> $ ./hack/pin-dependency.sh github.com/spf13/cobra 67fc4837d267bc9bfd6e47f77783fcc3dffc68de
> $ ./hack/update-vendor.sh
> ```
> 
> Please tell me if it's a wrong way.

@dtaniwaki You are doing in the right way.

What's missing here is you have to run `../hack/update-generated-docs.sh`. This should fix the failures here.

dtaniwaki commented:
It seems some of the dependencies are missing.

I can get the same error when I run `kubetest --build --up --test --down` in my local.

```
W0412 02:37:26.079] ERROR: /workspace/k8s.io/kubernetes/vendor/github.com/google/cadvisor/accelerators/BUILD:3:1: no such package 'vendor/github.com/mindprince/gonvml': BUILD file not found on package path and referenced by '//vendor/github.com/google/cadvisor/accelerators:go_default_library'
```

I'll add it with `./hack/pin-dependency.sh`.

dtaniwaki commented:
`github.com/mindprince/gonvml` is already in go.mod but it doesn't appear in vendor directory. I'll debug `hack/update-vendor.sh`.

dtaniwaki commented:
Found `github.com/mindprince/gonvml` is removed by `go mod tidy` because it's only used from packages in vendor. I'm investigating how to include it.

dtaniwaki commented:
I realized executing `./hack/update-vendor.sh` in the master branch of commit `ed2bdd53dc6f44f36f7912ed0e1a7e5ba800151b` also removes `github.com/mindprince/gonvml` from vendor. Could someone try it in your environment?

I use go v1.12.4.

dixudx commented:
> I realized executing `./hack/update-vendor.sh` in the master branch of commit `ed2bdd53dc6f44f36f7912ed0e1a7e5ba800151b` also removes `github.com/mindprince/gonvml` from vendor. Could someone try it in your environment?
> 
> I use go v1.12.4.

It works in go v1.12.4. Did not find package `github.com/mindprince/gonvml` get removed.

dtaniwaki commented:
I don't know what was happening when I tried it last week, but it seems I'm able to update vendor now.

dtaniwaki commented:
@dixudx I think it's ready now.

liggitt commented:
bumping cobra brought in a raft of other transitive bumps, which this didn't update:
github.com/cpuguy83/go-md2man v1.0.10
github.com/spf13/pflag v1.0.3
github.com/spf13/viper v1.3.2
gopkg.in/yaml.v2 v2.2.2

and a new dependency:
github.com/mitchellh/go-homedir v1.1.0

Rebasing on master, and running `hack/lint-dependencies.sh` will show which of these we are still pinned to old versions of.

dtaniwaki commented:
I tried to pin and update other dependencies but wasn't able to make it by the following error.

```
$ ./hack/update-vendor.sh
+++ [0417 13:30:41] logfile at /tmp/update-vendor.ywFZ/update-vendor.log
+++ [0417 13:30:41] go.mod: update staging references
+++ [0417 13:31:13] go.mod: propagate to staging modules
+++ [0417 13:31:18] go.mod: sorting staging modules
!!! Error in ./hack/update-vendor.sh:217
  Error in ./hack/update-vendor.sh:217. 'go list all >> "${LOG_FILE}" 2>&1' exited with status 1
Call stack:
  1: ./hack/update-vendor.sh:217 main(...)
Exiting with status 1
$ go list all
warning: ignoring symlink /Users/dtaniwaki/github/src/k8s.io/kubernetes/bazel-bin
warning: ignoring symlink /Users/dtaniwaki/github/src/k8s.io/kubernetes/bazel-genfiles
warning: ignoring symlink /Users/dtaniwaki/github/src/k8s.io/kubernetes/bazel-kubernetes
warning: ignoring symlink /Users/dtaniwaki/github/src/k8s.io/kubernetes/bazel-out
warning: ignoring symlink /Users/dtaniwaki/github/src/k8s.io/kubernetes/bazel-testlogs
warning: ignoring symlink /Users/dtaniwaki/github/src/k8s.io/kubernetes/cluster/gce/cos
warning: ignoring symlink /Users/dtaniwaki/github/src/k8s.io/kubernetes/cluster/gce/custom
warning: ignoring symlink /Users/dtaniwaki/github/src/k8s.io/kubernetes/cluster/gce/ubuntu
warning: ignoring symlink /Users/dtaniwaki/github/src/k8s.io/kubernetes/third_party/etcd
go: finding k8s.io/kubernetes/pkg/client/unversioned/remotecommand latest
go: finding k8s.io/kubernetes/pkg/client/clientset_generated/clientset/typed/core latest
go: finding k8s.io/kubernetes/pkg/client/clientset_generated/clientset latest
go: finding k8s.io/kubernetes/pkg/api latest
go: finding k8s.io/kubernetes/pkg/client/clientset_generated/clientset/fake latest
go: finding k8s.io/kubernetes/pkg/client/unversioned latest
go: finding k8s.io/kubernetes/pkg/client/clientset_generated/clientset/typed latest
go: finding k8s.io/kubernetes/pkg/client/clientset_generated latest
go: finding k8s.io/kubernetes/pkg latest
go: finding k8s.io/kubernetes/pkg/client latest
build k8s.io/kubernetes/cluster/images/etcd/migrate: cannot load github.com/ugorji/go/codec: ambiguous import: found github.com/ugorji/go/codec in multiple modules:
        github.com/ugorji/go v0.0.0-20171019201919-bdcc60b419d1 (/Users/dtaniwaki/.gvm/pkgsets/go1.12.4/kubernetes/pkg/mod/github.com/ugorji/go@v0.0.0-20171019201919-bdcc60b419d1/codec)
        github.com/ugorji/go/codec v0.0.0-20181204163529-d75b2dcb6bc8 (/Users/dtaniwaki/.gvm/pkgsets/go1.12.4/kubernetes/pkg/mod/github.com/ugorji/go/codec@v0.0.0-20181204163529-d75b2dcb6bc8)
```

It looks a similar issue to https://github.com/gin-gonic/gin/issues/1673.

dtaniwaki commented:
@liggitt 

I executed `hack/lint-dependencies.sh` and added the dependencies one by one, then more dependencies appeared when I added one and the commit logs of the corresponded updates became like below. The order is newest first.

```
866fa26de6 Update golang.org/x/net to v0.0.0-20190311183353-d88877176a
1bcb3e066a Update golang.org/x/tools to v0.0.0-20190328211700-ab21143f2384
5faa6c26b1 Update golang.org/x/sys to v0.0.0-20190312061237-fead79001313
adb8eaff60 Remove unused pinned module
295fa8b3e5 Update github.com/russross/blackfriday to v1.5.2
261d89e496 Update github.com/spf13/pflag to v1.0.3
a997ec49ab Update gopkg.in/yaml.v2 to v2.2.2
76d5fb9d4e Update github.com/onsi/ginkgo to v1.6.0
fb855b34d1 Update github.com/cpuguy83/go-md2man to v1.0.10
6dff2b18ce Update github.com/BurntSushi/toml to v0.3.1
```

Besides I can't update vendor for `github.com/spf13/viper v1.3.2` as I mentioned above, it looks I have to update so many files in vendor. Some of the dependencies are not mentioned in your comment above. Am I doing correct? I just want to update cobra version...

dtaniwaki commented:
I decided not to update cobra to the latest version of 67fc4837d267bc9bfd6e47f77783fcc3dffc68de as it requires viper to update too. I pinned the cobra version at ba1052d4cbce7aac421a96de820558f75199ccbc which includes my fix for two-word flags handling.

dtaniwaki commented:
@dixudx @liggitt I finally update vendor! Would you review the PR again?

soltysh commented:
@dtaniwaki please rebase your PR, and squash all the actual code changes into a single commit.

dtaniwaki commented:
@soltysh I rebased the PR and squashed my code changes. Could you review it?

dixudx commented:
@dtaniwaki Another rebase is needed.

dtaniwaki commented:
@dixudx Did it, thanks.

dixudx reviewed (approved):
/lgtm

dtaniwaki commented:
This seems endless.

liggitt commented:
`golang.org/x/{sync,sys,tools,net}` are currently on the go1.12 branch commits. no need to bump those.

dtaniwaki commented:
@liggitt So do you mean I can ignore `./hack/lint-dependencies.sh`?

liggitt commented:
> @liggitt So do you mean I can ignore the result of `./hack/lint-dependencies.sh`?

for the golang.org/x/... dependencies, yes. Opened https://github.com/kubernetes/kubernetes/pull/79944 to omit those from the output by default to avoid confusing people.

It looks like cobra tagged a release containing the fix we want, which is great. You can replace all the dependency bump commits with a single commit that just bumps cobra:

```
hack/pin-dependency.sh github.com/spf13/cobra v0.0.5
hack/update-vendor.sh
```

dtaniwaki commented:
The fix for two-word flags is [included in cobra v0.0.4](https://github.com/spf13/cobra/compare/v0.0.3...v0.0.4) which has been updated in https://github.com/kubernetes/kubernetes/pull/78187 already although it doesn't mention the fix.

dtaniwaki commented:
@liggitt As the fix for two-word flags has been included in cobra v0.0.4 which has been updated in Kubernetes, I renamed the title of this PR and committed only the change of custom func name and debug mode autocompletion. Could you review it?

liggitt commented:
/lgtm
/approve

tatsuhiro-t commented:
It looks like this PR breaks kubectl completion if override flags (e.g., --kubeconfig) are supplied.
With kubectl v1.16.0-alpha.2, `kubectl --kubeconfig=foo -n ` then hit TAB will show "specify resource type" error.
